fl4health/utils/dataset_converter.py

```diff
@@ -0,0 +1,159 @@
+        return self
+
+    def get_condition_vector_size(self) -> int:
+        if self.condition is None:
+            return 0
+        elif self.condition == "label":
+            assert self.dataset is not None, "Error: no dataset is passed to the converter."
+            return len(torch.unique(self.dataset.targets))
+        elif isinstance(self.condition, torch.Tensor):
+            return self.condition.size(0)
+        else:
+            raise NotImplementedError("Error: support for this type of condition is not added to the data converter.")
+
+    def _setup_converter_function(self) -> Callable:
+        """Sets the converter function for autoencoder based models (if it is not already specified by the user).
+        If condition is not None, the respective converter function accounts for the type of
+        condition and handles the concatenation.
+
+        Returns:
+            Callable: The suitable converter function based on the condition.
+        """
+        # If the autoencoder is conditional
+        if self.condition is not None:
+            if self.condition == "label":
+                # Condition depends on the target.
+                converter_function = self._cat_input_label
+            elif isinstance(self.condition, torch.Tensor):
+                converter_function = self._cat_input_condition
+            else:
+                raise NotImplementedError("Error: support for this type of condition is not added.")
+
+        elif self.condition is None:  # non-conditional autoencoder
+            converter_function = self._only_replace_target_with_data
+        return converter_function
+
+    def _only_replace_target_with_data(
+        self, data: torch.Tensor, target: Union[None, torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """The data converter function used for simple autoencoders or variational autoencoders."""
+        # Target in self-supervised training for autoencoder is the data.
+        return data, data
+
+    def _cat_input_condition(
+        self, data: torch.Tensor, target: Union[None, torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """The data converter function used for conditional autoencoders.
+        This converter is used when we have a torch tensor as condition for all the data samples.
+        """
+        # We can flatten the data since self.data_shape is already saved.
+        # Target should be the original data.
+        assert isinstance(self.condition, torch.Tensor), "Error: condition should be a torch tensor"
+        return torch.cat([data.view(-1), self.condition]), data
+
+    def _cat_input_label(self, data: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """The data converter function used for conditional autoencoders.
+        This converter is used when we want to condition each data sample on its label.
+        """
+        num_conditions = self.get_condition_vector_size()
```

Comment:
There is an implicit assumption that the target tensor here is a class label. That is, the target tensor is a 1D tensor with the size of a batch. However, it's possible that the target will have already been processed into a batch of one-hot tensor. There is also the possibility that the target isn't supposed to be one-hot encoded but is rather a multi-class label or a "dense" label. I think we want to support these as well based on the tensor shape and type?

Reply:
We could simply have an option in the constructor that specifies whether to one-hot encode the labels or leave as is as well.